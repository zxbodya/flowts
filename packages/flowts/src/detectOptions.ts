import * as babel from '@babel/core';
import traverse from '@babel/traverse';
import { sharedParserPlugins } from './sharedParserPlugins';

export interface SourceOptions {
  isJSX: boolean;
  isFlow: boolean;
}

export function detectOptions(source: string, filename: string): SourceOptions {
  const flowAst = babel.parseSync(source, {
    compact: false,
    babelrc: false,
    configFile: false,
    ast: true,
    parserOpts: {
      allowReturnOutsideFunction: true,
      plugins: ['flow', 'jsx', ...sharedParserPlugins],
    },
    filename,
  });

  let isJSX = /\.jsx$/i.test(filename);
  let isFlow = /@flow/.test(source) || /\.js\.flow|\.flow\.js$/i.test(filename);

  if (flowAst === null) {
    throw new Error(
      'babel.parseSync returned null - probably there is some configuration error'
    );
  }

  // workaround to have "scope.hub"
  // @ts-ignore
  new babel.File({ filename }, { code: source, ast: flowAst });

  traverse(flowAst, {
    JSX() {
      isJSX = true;
    },
    Flow() {
      isFlow = true;
    },
  });
  return { isJSX, isFlow };
}
