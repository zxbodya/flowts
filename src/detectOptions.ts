import * as babel from '@babel/core';
import traverse from '@babel/traverse';
import { sharedParserPlugins } from './sharedParserPlugins';

export interface SourceOptions {
  isJSX: boolean;
  isFlow: boolean;
}

export function detectOptions(source: string, filename: string): SourceOptions {
  const flowAst = babel.parseSync(source, {
    babelrc: false,
    ast: true,
    parserOpts: {
      plugins: ['flow', 'jsx', ...sharedParserPlugins],
    },
    filename,
  });

  let isJSX = /\.(jsx|tsx)$/i.test(filename);
  let isFlow = /@flow/.test(source) || /\.js\.flow$/i.test(filename);

  if (flowAst === null) {
    throw new Error(
      'babel.parseSync returned null - probably there is some configuration error'
    );
  }

  traverse(flowAst, {
    JSX() {
      isJSX = true;
    },
    Flow() {
      isFlow = true;
    },
  });
  return { isJSX, isFlow };
}
