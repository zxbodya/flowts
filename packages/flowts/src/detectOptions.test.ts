import { detectOptions } from './detectOptions';

describe('detectOptions', () => {
  it('detects @flow comment', () => {
    expect(detectOptions(`// @flow`, 'flow.js')).toMatchSnapshot();
  });
  it('detects flow constructions in code', () => {
    expect(detectOptions(`let a: number = 1;`, 'test.js')).toMatchSnapshot();
  });
  it('detects JSX', () => {
    expect(detectOptions(`let a = <div/>;`, 'test.js')).toMatchSnapshot();
  });
  it('throws', () => {
    expect(() => detectOptions(`let a = /<div/>;`, 'test.js')).toThrow();
  });
  it('use file extension to detect jsx', () => {
    expect(detectOptions(`// @flow`, 'flow.jsx')).toMatchSnapshot();
  });
  it('use file extension to detect flow', () => {
    expect(detectOptions(`let a = 1;`, 'test.js.flow')).toMatchSnapshot();
    expect(detectOptions(`let a = 1;`, 'test.flow.js')).toMatchSnapshot();
  });
});
