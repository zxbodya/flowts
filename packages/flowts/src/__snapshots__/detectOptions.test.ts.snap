// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detectOptions detects @flow comment 1`] = `
Object {
  "isFlow": true,
  "isJSX": false,
}
`;

exports[`detectOptions detects JSX 1`] = `
Object {
  "isFlow": false,
  "isJSX": true,
}
`;

exports[`detectOptions detects flow constructions in code 1`] = `
Object {
  "isFlow": true,
  "isJSX": false,
}
`;

exports[`detectOptions use file extension to detect flow 1`] = `
Object {
  "isFlow": true,
  "isJSX": false,
}
`;

exports[`detectOptions use file extension to detect flow 2`] = `
Object {
  "isFlow": true,
  "isJSX": false,
}
`;

exports[`detectOptions use file extension to detect jsx 1`] = `
Object {
  "isFlow": true,
  "isJSX": true,
}
`;
